File: 0118-pascals-triangle/0118-pascals-triangle.js
/**
 * @param {number} numRows
 * @return {number[][]}
 */
var generate = function(numRows) {
    if (numRows === 0) {
        return [];
    }

    let result = [[1]];

    for (let i = 1; i < numRows; i++) {
        let prevRow = result[i - 1];
        let newRow = new Array(i + 1);
        newRow[0] = 1;
        newRow[i] = 1;

        for (let j = 1; j < i; j++) {
            newRow[j] = prevRow[j - 1] + prevRow[j];
        }

        result.push(newRow);
    }

    return result;
};

File: 0118-pascals-triangle/0118-pascals-triangle.java

class Solution {
    public List<List<Integer>> generate(int numRows) {
        if(numRows <= 0) return new ArrayList<>();
        
        List<List<Integer>> triangle = new ArrayList<>();
        buildTriangle(triangle, numRows, 0);
        return triangle;
    }
    
    private void buildTriangle(List<List<Integer>> triangle, int totalRows, int currentRow) {
        // Base case
        if(currentRow >= totalRows) return;
        
        // Build current row
        List<Integer> row = new ArrayList<>();
        for(int i = 0; i <= currentRow; i++) {
            if(i == 0 || i == currentRow) {
                row.add(1);
            } else {
                // Get previous row and sum adjacent elements
                List<Integer> prevRow = triangle.get(currentRow - 1);
                row.add(prevRow.get(i - 1) + prevRow.get(i));
            }
        }
        
        triangle.add(row);
        
        // Recursive call for next row
        buildTriangle(triangle, totalRows, currentRow + 1);
    }
}